bft/bft_types/src/lib.rs

```diff
@@ -89,6 +90,137 @@ impl fmt::Display for PositionedInstruction {
         write!(f, "{}:{} {}", self.line, self.character, self.instruction)
     }
 }
+/// Instructions that have been processed into a form useful to an interpreter
+pub enum DecoratedInstruction {
+    /// A loop has been opened. In addition, here is where it will close
+    OpenLoop {
+        instruction: PositionedInstruction,
+        closer: PositionedInstruction,
+        closer_index: usize,
+    },
+    /// A loop has been closed. In addition, here is where it was opened
+    CloseLoop {
+        instruction: PositionedInstruction,
+        opener: PositionedInstruction,
+        opener_index: usize,
+    },
+    /// An ordinary instruction that can be used as-is
+    Instruction(PositionedInstruction),
+    /// An open-bracket instruction that gets replaced when the loop is closed
+    ///
+    /// This is expected to only exist as a temporary implementation detail.
+    PlaceholderOpenBracket,
+}
+
+/// A program that's been processed into a form useful to an interpreter
+/// Compared to a Program, this has the additional constraint that the code must be valid Brainfuck.
+pub struct DecoratedProgram {
```

Comment:
I like the idea that you've gone for using types to enforce the validity invariant.